### ISSUE TITLE
Hard Timeout

### ISSUE THREAD

**Issue by jamesgraves** [jamesgraves/bonzAI#4]
So this just happened:

    Script execution has been terminated with a hard reset: CPU limit reached

    Your script is temporary blocked due to a hard reset inflicted to the runtime process.
    Please try to change your code in order to prevent causing hard timeout resets.

**Comment by jamesgraves** [comment]
Further details:

Single room.
User Global Control Level (GCL): 2
Room Control Level (RCL): 5
Should have been about 6 or so creeps running at the time.

**Comment by bonzaiferroni** [comment]
yeah, this is the bane of my existence right now! I'm having a hard time figuring out where they are coming from. So you started getting these when you started using the testing branch?

**Comment by jamesgraves** [comment]
This was a while back, running on the `master` branch.

I've been thinking about restructuring how the code is scheduled.  For example, with the IgorMission we have this for finding the idle position:

```typescript
private findIgorIdlePosition() {
    if (!this.memory.idlePosition && Game.time % 1000 === 0) {
```

So this will run the given function once every thousand ticks if there is no idle position, because it shouldn't need to change very often.  That's fine by itself.

However, how many other things are currently running at tick 1000, 2000, ..., 31000, 555000, and so on?  Maybe too many?  Well, we can search through the code to find out.  But the current structure makes it harder to get a global view of what's running when.

I'm currently thinking about a set of nested `switch` statements more like this:

```typescript
switch (Game.time % 10) {
    case 1:
        // task that needs to run once every 10 ticks.
        break;
    case 2:
        // another task that needs to run once every 10 ticks.
        break;


    case 0:
        switch (Game.time % 100) {
            case 10:
                // task that needs to run once every 100 ticks.
                break;
            case 20:
                // another task that needs to run once every 100 ticks.
                break;

                // other tasks...

            case 0:
                switch (Game.time % 1000) {
                    case 100:
                        // task that needs to run once every 1000 ticks.
                        break;
                }
                break;
           }
           break;
}
```

The point of this is to ensure that only one occasional task is running at a time, and really spread out the load.

I realize this is in some ways the very antithesis of object-oriented programming.  But due to the tight execution time constraints, I am considering it.

Just something to think about... I don't have any concrete plans to try this out just yet.

**Comment by bonzaiferroni** [comment]
Yeah, the `Game.time % 1000 === 0` is a pretty clumsy way to handle task scheduling, and I only used it in places where the code shouldn't actually make much of a difference in cpu. I've started using a pattern like this:

```
if (Game.time < memory.nextCheck) { return; }
// task
memory.nextCheck = Game.time + helper.randomInterval(1000)
```

The helper.randomInterval(1000) function returns some integer near the value passed as an argument, +/- 10%. This works a lot better. 

Yeah, this isn't an easy problem to solve with OOP considering the lifespan of our objects is only a single game tick. Still, I bet there is a more elegant way to handle it. A good solution needs to have at least the following:

1. Very little overhead - the "nextCheck" pattern above has just a boolean evaluation.
2. Low memory footprint - "nextCheck" only needs a single stored number value. Still, the original solution doesn't need any memory at all. 
3. Intuitive and cohesive - "nextCheck" is not very cohesive, you have the boolean check, the value assignment, and the memory variable declaration in three different places

I also toyed with running code based on probability: 
```
if (Math.random > .001) { return }
// some task that runs every 1000 ticks on average
```

I wasn't sure how much cpu Math.random uses, I'm sure it is negligible. The main "problem" was the lack of reliability, there is some chance that code that needs to run at least every x ticks would be significantly delayed.

**Comment by bonzaiferroni** [comment]
on the subject of hard resets, it actually doesn't seem likely that it is a task scheduling issue, since this is more likely to trigger a "soft reset", at least according to my understanding of the discussion here: http://support.screeps.com/hc/en-us/community/posts/115000235905-Clarification-request-on-Cpu-limit-abusing-

According to that, the most likely culprit is "running a heavy native function like JSON.parse, or some types of infinite loops".

In my experiments, I wasn't able to trigger a hard reset with an infinite loop (always resulted in an soft reset) or infinite recursion (always triggered a max call stack reached exception). The only way I was able to trigger a hard reset was making a humongous object and calling JSON.stringify.

My next course of investigation will involve just taking sections out of my main loop to see if it stops the resets. It is actually really helpful to have your report that you are getting them to. I haven't yet seen whether they get triggered on a private server or not. This would be a great way to debug them.

**Comment by bonzaiferroni** [comment]
One other thing, it seems as though everyone gets hard resets from time to time, it is pretty likely that most of these are due to server issues and not your code. Some people even report them occasionally with an empty loop.

I think my resets might be related to my code though, since I get them on a pretty steady basis. Other people report getting like a dozen within 24 hours and then none for several days, whereas I get them every day.

**Comment by jamesgraves** [comment]
I did just receive a 1-day token for recent server outages, so maybe that was the original cause of the hard timeout.

I was initially suggesting a fixed schedule using nested switch statements because this has minimal compute time overhead and no storage.  But it doesn't give you any flexibility either.  For example, if the room is under attack, you'd want to run certain tasks more frequently than normal, and then go back to the regular schedule when the attack is over.

One of my main worries with the nextCheck pattern is there is still a call / return overhead as well for each task.  But I haven't done any profiling to see if this is actually a concern.  At any rate, what's nice is that each task can flexibly schedule itself.

If this was a "normal" runtime environment with persistent memory, I'd be looking at implementing a time-based priority queue.  This is where you set a timeout and a priority for each task, and they bubble up to the head as time expires.  The scheduler can then execute as many or as few as possible based on the available CPU time.  It can also decide to run some tasks a little early if there is a surplus of CPU time currently available.

As with nextCheck, each task then schedules itself to run again later.

However, since this is a more complicated data structure, the penalty for serialize / deserialize may be impractical.  I'm not sure.  Maybe this will be too complicated.

I haven't set up a private server yet, but that is on the TODO list as well.

**Comment by jamesgraves** [comment]
Just got another hard timeout error this morning.

**Comment by bonzaiferroni** [comment]
I'm getting about a dozen a day here. Almost everyone gets them, so the occasional hard reset might not even be related to your code. I think I get them at an above normal rate, so I'm still trying to see if I can find the cause. 

I just started running the private server, so I'll see if I get the timeouts there too.

**Comment by jamesgraves** [comment]
I just got a private server running too on an old laptop.  It is fun to watch the initial base build because the ticks are faster than on the main server.

I'm not super-worried about these hard timeouts, but I'll keep this issue open for now.

**Comment by bonzaiferroni** [comment]
I'm trying out a new way to schedule tasks. It is nothing fancy at all, but I think it is better than the "nextCheck" pattern. Really, it is just encapsulating that pattern so that it is all in one place. 

So rather than this:
```
if (Game.time < memory.nextCheck) { return; }
memory.nextCheck = Game.time + helper.randomInterval(1000)
// task
```

You can do this: 
``` 
if (Scheduler.delay(this, "nextCheck", 1000)) { return; }
// task
```

Not too different but it makes it more standard. Also, this is intended to be used with any code that can be delayed, so it won't bother running the code if cpu is already too high on that tick. All the code is here, very short:

https://github.com/bonzaiferroni/bonzAI/blob/testing/src/Scheduler.ts

**Comment by bonzaiferroni** [comment]
While I was implementing it in all the relevant places, I found some code that could very well be causing hard timeouts. It was the findProcess() function in IgorMission. It wasn't clear, but this could could run all on the same tick, and it is potentially very expensive. Hopefully I see fewer timeouts as a result of fixing that.

**Comment by jamesgraves** [comment]
Great! I'll take a look at your changes, and keep an eye out for anything else that can be causing a problem.  I haven't seen any hard timeouts yet on my private server.  I need to get email notifications set up for that too.
